File: src/main/java/com/example/chtl/dispatch/CompilerDispatcher.java
package com.example.chtl.dispatch;

import com.example.chtl.compilers.ChtlCompiler;
import com.example.chtl.compilers.ChtlJsCompiler;
import com.example.chtl.compilers.CssCompiler;
import com.example.chtl.compilers.JsCompiler;
import com.example.chtl.core.Fragment;
import com.example.chtl.core.FragmentType;
import com.example.chtl.core.ScanResult;

public class CompilerDispatcher {
	private final ChtlCompiler chtlCompiler = new ChtlCompiler();
	private final ChtlJsCompiler chtlJsCompiler = new ChtlJsCompiler();
	private final CssCompiler cssCompiler = new CssCompiler();
	private final JsCompiler jsCompiler = new JsCompiler();

	public CompilationResult dispatch(ScanResult scanResult) {
		// CHTL 主流程
		ChtlCompiler.Output chtlOut = chtlCompiler.compile(scanResult.getSource());

		StringBuilder globalCss = new StringBuilder();
		StringBuilder globalJs = new StringBuilder();

		// CHTL 输出的全局产物
		if (!chtlOut.emittedGlobalCss.isEmpty()) globalCss.append(chtlOut.emittedGlobalCss).append('\n');
		if (!chtlOut.emittedGlobalJs.isEmpty()) globalJs.append(chtlOut.emittedGlobalJs).append('\n');

		// 收集 Origin CSS 与 全局 CSS（来自局部style中产生的）
		for (Fragment f : scanResult.getFragmentsByType(FragmentType.CSS)) {
			String compiled = cssCompiler.compile(f.content());
			if (!compiled.isBlank()) {
				globalCss.append(compiled).append('\n');
			}
		}

		// 收集脚本：普通 JS + CHTL JS 转译为 JS
		for (Fragment f : scanResult.getFragmentsByType(FragmentType.JS)) {
			String compiled = jsCompiler.compile(f.content());
			if (!compiled.isBlank()) globalJs.append(compiled).append('\n');
		}
		for (Fragment f : scanResult.getFragmentsByType(FragmentType.CHTL_JS)) {
			String transpiled = chtlJsCompiler.compile(f.content());
			if (!transpiled.isBlank()) globalJs.append(transpiled).append('\n');
		}

		return new CompilationResult(chtlOut.htmlBody, globalCss.toString(), globalJs.toString());
	}
}

File: chtl_compiler/dispatcher/compiler_dispatcher.py
from __future__ import annotations

from typing import List

from chtl_compiler.types import CompilationBundle, CompilationUnit, Fragment, FragmentKind
from chtl_compiler.compilers.chtl_compiler import ChtlCompiler
from chtl_compiler.compilers.chtl_js_compiler import ChtlJsCompiler
from chtl_compiler.compilers.css_compiler import CssCompiler
from chtl_compiler.compilers.js_compiler import JsCompiler


class CompilerDispatcher:
	def __init__(self) -> None:
		self._chtl = ChtlCompiler()
		self._chtl_js = ChtlJsCompiler()
		self._css = CssCompiler()
		self._js = JsCompiler()

	def dispatch(self, fragments: List[Fragment]) -> CompilationBundle:
		units: List[CompilationUnit] = []
		for fragment in fragments:
			if fragment.kind == FragmentKind.CHTL:
				units.append(self._chtl.compile(fragment.content))
			elif fragment.kind == FragmentKind.CHTL_JS:
				units.append(self._chtl_js.compile(fragment.content))
			elif fragment.kind == FragmentKind.CSS:
				units.append(self._css.compile(fragment.content))
			elif fragment.kind == FragmentKind.JS:
				units.append(self._js.compile(fragment.content))
		return CompilationBundle(units=units)